daffodil-io/src/main/scala/org/apache/daffodil/io/InputSource.scala

```diff
@@ -340,6 +352,9 @@ class BucketingInputSource(inputStream: java.io.InputStream, bucketSize: Int = 1
       while (bytesStillToGet > 0) {
         val bytesToGetFromCurrentBucket = Math.min(bucketSize - byteIndex, bytesStillToGet).toInt
 
+        if (buckets(bucketIndex.toInt) == null)
```

Comment:
Interesting point. If one call to parse runs into this bucket overflow situation, they "ruin" the input stream for subsequent calls to parse, should the parse return in error. We can't unwind back to the starting position for that call to parse. We've lost that location. 

Clearly a Runtime SDE is fatal, and not recoverable, but it indicates more than just a failure of that call to parse. Rather,  the whole input stream is also unusable now, unless we can somehow know that it was successfully restored to where it started with the parse call.  The first top-level call to parse for a data input stream could save the current position. On unwind from that parse call, we could check if the stream has been restored to that position (or can be restored to that position - and restore it to that when possible), and if not mark it as in some "broken due to pre-advance by failed parse" state. Then an attempt to reuse that data input stream would throw indicating why the stream is unsuitable for use.

Reply:
I think my most recent commit addresses this with the added check on Pstate.isDataStreamValid, but I'm not sure I understand how this situation works where the user initiates another parse with the same data input stream, so I don't have any tests to cover this.

Reply:
I think you'd have to write some unit test code, and maybe a CLI test using the ``--stream`` option would be good too.

Essentially, you need data + a schema that will fail due to a backtracking exception. But you need to continue to try to parse that. I *think* the CLI --stream option will abort the parse and won't try again, so it would be good to have that CLI tests. A unit test would look something like

```
val input = // create an InputSourceDataInputStream()
val output = // create infoset outputter

val pr1 = dataProcessor.parse(input, output) // should cause an backtracking SDE result
val pr2 = dataProcessor.parse(input, output) // should cause an invalid usage error
```
So we try to parse the data once, that fails and makes our input source invalid. Then we try to parse again using the same input source, but that should fail since the input is not invalid.

Thinking about this, this doesn't even need to be a backtracking SDE. Any runtime SDE should work.